doc/languages-frameworks/agda.section.md

```diff
@@ -0,0 +1,85 @@
+---
+title: Agda
+author: Alex Rice (alexarice)
+date: 2020-01-06
+---
+# Agda
+
+## How to use Agda
+
+Agda can be installed with `agda.Agda`. For example:
+```
+$ nix-env -iA agda.Agda
+```
+
+To use agda with libraries, the `withPackages` function can be used. This function either takes:
++ A list of packages,
++ or a function which returns a list of packages when given the `agda` attribute set,
++ or an attribute set containing a list of packages and local home libraries (see below).
+
+For example, suppose we wanted a version of agda which has access to the standard library. This can be obtained with the expressions:
+
+```
+agda.withPackages [ agda.standard-library ]
+```
+
+or
+
+```
+agda.withPackages (p: [ p.standard-library ])
+```
+
+If you want to use a library in your home directory (for instance if it doesn't have a nix derivation or is a development version) then an attribute set can be passed into `agda.withPackages` with the following arguments:
+
+```
+agda.withPackages {
+  pkgs = [ agda.standard-library ];
+  homeLibraries = "path/to/my/package.agda-lib";
+}
+```
+Multiple libraries can be passed in by separating them with newlines.
+
+Agda will not by default use these libraries. To tell agda to use the library we have some options:
+- Call `agda` with the library flag:
+```
+$ agda -l standard-library -i . MyFile.agda
+```
+- Write a `my-library.agda-lib` file for the project you are working on which may look like:
+```
+name: my-library
+include: .
+depends: standard-library
+```
+- Create the file `~/.agda/defaults` and add any libraries you want to use by default.
+
+More information can be found [here](https://agda.readthedocs.io/en/v2.6.0.1/tools/package-system.html).
+
+## Compiling Agda
+Agda modules can be compiled with the `--compile` flag. A version of `ghc` with `ieee` is made available to the Agda program via the `--with-compiler` flag.
+This can be overridden by a different version of `ghc` as follows:
+
+```
+agda.withPackages {
+  pkgs = [ ... ];
+  ghc = myGhcDerivation;
+}
+```
+
+## Writing Agda packages
+To write a nix derivation for an agda library, first check that the library has a `*.agda-lib` file.
+
+A derivation can then be written using `agda.mkDerivation`. This has similar arguments to `stdenv.mkDerivation` with the following exceptions:
++ The `buildInputsAgda` should be used for agda library dependencies.
```

Comment:
I'm not sure there is a way to recognise agda libraries at the moment, could add a flag like the old builder had. I'm not really sure on the naming, would `agdaBuildInputs` be more in line with the rest of nixpkgs

Reply:
Let's consider typical options for `buildInputs`. You could do it during evaluation by evaluating a predicate like `(p: p ? everythingFile)`, or you could add a `setupHook` in `Agda.mkDerivation` in which Agda derivations will populate some shell variable (":" separated string or a bash array). Latter option should allow a single derivation to potentially provide several libraries.

Reply:
I don't understand the `setupHook` thing at all I'm afraid. Are you suggesting populating some shell variable and then trying to get agda to use that shell variable when building? I'm not sure this will work

Reply:
Well, if agda doesn't accept environment options, then your wrapper could, or the builder for the wrapper could.

I believe there, more popular option to only have the hook in the main Agda derivation. I believe `addEnvHooks` (https://nixos.org/nixpkgs/manual/#ssec-setup-hooks) should run it over `buildInputs` for you:
* https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/setup-hook.sh
* https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/perl/setup-hook.sh
* https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/typesetting/tex/texlive/setup-hook.sh - example with a check (for the path you can always use "$out/nix-support/agda" if you are out of ideas)

Reply:
You also get `propagatedBuildInputs` working for free as well as some cross stuff (may not be so relevant for Agda).

Reply:
I have very little knowledge about how this corner of nix works but I can take a look (probably tomorrow)

Reply:
I'm not sure if this is feasable. I assume the effects of setup hooks are only available in the actual build script, and so I wouldn't be able to use nix to build the library-file if I was using setup hooks to collect the dependencies?

Reply:
If you want to build a library file I believe the best way is to add an attribute `isAgdaDerivation` and filter on it on all `buildInputs`. The setup hooks way only works if Agda can get all its packages from an environment variable (without a libraries file).

Reply:
I feel this is the cleanest way to do it

Reply:
Seems you have to update this doc line still?